PROTOCOL.md

```diff
@@ -0,0 +1,127 @@
+data storage.
+
+For one particular operation the read took 4.1 milliseconds. You have
+chosen to use "duration" as the measurement name for anything involving
+execution time. In that case, the StatsD message payload for the
+roundtrip time taken could be:
+
+  - `duration:4.1|ms|#service=login,team=myteam,operation=read\n`
+
+In order to track the status of their service, the team also records
+each returned status code. For a service with a non-trivial amount of
+activity, it is important to know how many times any particular status
+code has been served. So, just before returning the service response,
+the service could send the following message to bucky3:
+
+  - `status:200|h|#service=login,team=myteam,route=/user/login\n`
+
+Of course, if a login was unsuccesful, the message could instead be:
+
+  - `status:401|h|#service=login,team=myteam,route=/user/login\n`
+
+Another team maintains a service with persistent real-time connections.
+In addition to response status codes, they care roundtrip times and the
+number of connected clients. Every couple of seconds, they record the
+number of currently open client connections, and at a particular moment
+they had 473 connections. The message payload could then be:
+
+  - `connections:473|g|#service=ourstream,team=otherteam\n`
+
+In addition, every time they serve a request, they record both the
+response status and the roundtrip time. So on every succesful response
+they could send the following two messages:
+
+- `status:200|h|#service=ourstream,team=otherteam,action=something\n`
+- `duration:2.9|ms|#service=ourstream,team=otherteam,action=something\n`
+
+### The Edge Case
+
+So far the message format has been very easy to follow, but of course
+there is a special case.
+
+Practically every service needs to record how many requests it handles
+within any given time unit. The message format for counter is slightly
+different because it uses a fixed value. Counter is internally
+maintained by bucky3, and gets reset on every measurement window change.
+A counter increase simply raises the current value by 1. A message to
+increase a counter would look something like:
+
+  - `requests:1|c|#service=myservice,team=someteam,route=/some/path\n`
+
+The counter values are most often used as sums over a time window for
+measuring requests per second (RPS) or requests per minute (RPM).
+
+Now, *technically* a counter can be increased by any amount, so the
+value after the semicolon may be other than 1. Bucky3 does handle this
+case, making it possible to nudge a counter up by arbitrary amounts.
+However, in practice you should never need to do this. For sanity,
+simply increase a counter always by 1.
+
+(There are some very specific high-frequency environments, where
+overriding the counter makes sense. Even then, it is recommended that
+you use a gauge instead. Flooding bucky3 with extremely rapid
+counter-increase messages has the potential to degrade monitoring
+performance.)
```

Comment:
The rate option in protocol helps in such cases, too. Counter is only type of aggregate where the use of rate actually makes clear sense.